### DISCUSSION TITLE
FYI. msquic-async crate

### DISCUSSION THREAD

**Discussion by masa-koz** [microsoft/msquic#4707] [Show and tell]
I created `msquic-async` MsQuic based quic library that supports async operation for rust.
https://github.com/masa-koz/msquic-async-rs

Using this crate, you can use MsQuic as below.
```rust
    let mut stream = conn
        .open_outbound_stream(msquic_async::StreamType::Bidirectional, false)
        .await?;
    stream.write_all("hello".as_bytes()).await?;
    let mut buf = [0u8; 1024];
    let len = stream.read(&mut buf).await?;
    info!("received: {}", String::from_utf8_lossy(&buf[0..len]));
```

Regards,

**Comment by nibanks** [comment]
I haven't looked at the details of the crate. How do you handle buffers on the send and receive path? Do you use send buffering (i.e. copies data)? On the receive path, do you return pending or do you copy the buffers somewhere?

**Reply by masa-koz** [reply]
On the send path, `Stream::poll_write()` copies data. But `Stream::zerocopy_write()` accepts a `bytes::Bytes` and pass it to MsQuic directly.
On the recieve path, msquic-async always stores the buffer and returns QUIC_STATUS_PENDING in the handler of STREAM_EVENT_RECEIVE. `Stream::poll_read()` takes the buffer and copies the data into user's buffer. `Stream::poll_read_chunk()` takes the buffer and return it. The buffer, `StreamRecvBuffer` implements `bytes::Buf`, so user can process it.

**Comment by masa-koz** [comment]
I've published this in crates.io.
https://crates.io/crates/msquic-async

**Reply by masa-koz** [reply]
I published [h3-msquic-async](https://crates.io/crates/h3-msquic-async) to use MsQuic with [h3](https://github.com/[hyperium/h3](https://github.com/hyperium/h3)).
